File: typescript/processor/src/reader/description-inferrer.ts
// @docspec:module {
//   id: "docspec-ts-description-inferrer",
//   name: "Description Inferrer",
//   description: "Generates human-readable descriptions from camelCase identifiers by splitting on uppercase boundaries and mapping the leading verb to a natural-language prefix (e.g., 'getUserProfile' becomes 'Retrieves user profile.'). Used as the Tier 0 fallback when no TSDoc or decorator descriptions are present.",
//   since: "3.0.0"
// }

/**
 * @docspec:boundary "Verb-based description generation from identifier names"
 * @docspec:deterministic
 * @docspec:intentional "Converts method/class names into human-readable documentation strings"
 */
export class DescriptionInferrer {
  private static readonly VERB_MAP: Record<string, string> = {
    get: "Retrieves", find: "Finds", fetch: "Fetches", load: "Loads", read: "Reads",
    create: "Creates", add: "Adds", insert: "Inserts", save: "Saves", store: "Stores",
    update: "Updates", modify: "Modifies", patch: "Patches", change: "Changes",
    delete: "Deletes", remove: "Removes", destroy: "Destroys", drop: "Drops",
    validate: "Validates", check: "Checks", verify: "Verifies", assert: "Asserts",
    process: "Processes", handle: "Handles", execute: "Executes", run: "Runs",
    send: "Sends", publish: "Publishes", emit: "Emits", dispatch: "Dispatches",
    convert: "Converts", transform: "Transforms", map: "Maps", parse: "Parses",
    calculate: "Calculates", compute: "Computes", count: "Counts", sum: "Sums",
    is: "Checks whether", has: "Checks if it has", can: "Determines if it can",
    should: "Determines whether it should",
  };

  /** @docspec:deterministic */
  infer(name: string): string {
    // Split camelCase into words
    const words = name.replace(/([A-Z])/g, " $1").trim().split(/\s+/);
    if (words.length === 0) return "";

    const verb = words[0].toLowerCase();
    const rest = words.slice(1).map(w => w.toLowerCase()).join(" ");
    const prefix = DescriptionInferrer.VERB_MAP[verb];

    if (prefix && rest) {
      return `${prefix} ${rest}.`;
    }
    if (prefix) {
      return `${prefix} the resource.`;
    }

    // Fallback: humanize the name
    return words.join(" ").replace(/^\w/, c => c.toUpperCase()) + ".";
  }
}

File: packages/processor/python/docspec_processor/reader/description_inferrer.py
"""Infer descriptions from names."""
# @docspec:module {
#   id: "docspec-py-description-inferrer",
#   name: "Description Inferrer",
#   description: "Infers human-readable descriptions from class and function names by splitting camelCase/snake_case and mapping verb prefixes to natural language.",
#   since: "3.0.0"
# }
import re

VERB_MAP = {
    "get": "Retrieves", "find": "Finds", "fetch": "Fetches", "load": "Loads",
    "create": "Creates", "add": "Adds", "insert": "Inserts", "save": "Saves",
    "update": "Updates", "modify": "Modifies", "patch": "Patches",
    "delete": "Deletes", "remove": "Removes", "destroy": "Destroys",
    "validate": "Validates", "check": "Checks", "verify": "Verifies",
    "process": "Processes", "handle": "Handles", "execute": "Executes",
    "send": "Sends", "publish": "Publishes", "emit": "Emits",
    "convert": "Converts", "transform": "Transforms", "parse": "Parses",
    "calculate": "Calculates", "compute": "Computes", "count": "Counts",
    "is": "Checks whether", "has": "Checks if has", "can": "Determines if can",
}


# @docspec:boundary "Name-to-description inference for undocumented elements"
class DescriptionInferrer:
    # @docspec:method { since: "3.0.0" }
    # @docspec:deterministic
    # @docspec:intentional "Generate a human-readable description from a class or function name"
    # @docspec:preserves "Always returns a non-empty string ending with a period"
    def infer(self, name: str) -> str:
        words = re.sub(r"([A-Z])", r" \1", name).replace("_", " ").strip().split()
        if not words:
            return ""
        verb = words[0].lower()
        rest = " ".join(w.lower() for w in words[1:])
        prefix = VERB_MAP.get(verb)
        if prefix and rest:
            return f"{prefix} {rest}."
        if prefix:
            return f"{prefix} the resource."
        return " ".join(words).capitalize() + "."
